crates/polars-plan/src/logical_plan/optimizer/cluster_with_columns.rs

```diff
@@ -0,0 +1,561 @@
+                input_wc
+                    .exprs
+                    .exprs_mut()
+                    .extend(std::mem::take(current_wc.exprs.exprs_mut()));
+                std::mem::swap(current_wc.exprs.exprs_mut(), input_wc.exprs.exprs_mut());
+
+                *current_wc.input = *input_wc.input;
+                // @TODO: Is this allowed?
+                *current_wc.options = *input_wc.options;
+
+                self.lp_arena.take(input);
+
+                // Since we merged the current and input nodes and the input node might have
+                // optimizations with their input, we loop again on this node.
+                continue;
+            }
+
+            let mut current_schema = current_wc.schema.as_ref().clone();
+            let mut input_schema = input_wc.schema.as_ref().clone();
+
+            // @NOTE: We don't have to insert a SimpleProjection or redo the `current_schema` if
+            // `pushable` contains only 0..N for some N.
+            let mut current_exprs = std::mem::take(current_wc.exprs.exprs_mut())
+                .into_iter()
+                .zip(pushable)
+                .enumerate()
+                .filter_map(|(i, (expr, do_pushdown))| {
+                    if do_pushdown {
+                        // @NOTE: It would be nice to have a `remove_at_index` here.
+                        let (column, _) = current_wc.schema.get_at_index(i).unwrap();
+
+                        input_wc.exprs.exprs_mut().push(expr);
+                        let datatype = current_schema.remove(column).unwrap();
+                        input_schema.with_column(column.clone(), datatype);
+
+                        None
+                    } else {
+                        Some(expr)
+                    }
+                })
+                .collect();
+
+            std::mem::swap(&mut current_exprs, current_wc.exprs.exprs_mut());
+
+            current_schema.merge(input_schema.clone());
+
+            let proj_schema = current_wc.schema.clone();
+
+            *current_wc.schema = Arc::new(current_schema);
+            *input_wc.schema = Arc::new(input_schema);
+
+            // @NOTE: Here we add a simple projection to make sure that the output still
+            // has the right schema.
+            // @TODO: I don't think we always have to add a simple projection, maybe we can
+            // filter out some of the cases where we don't have to add it.
+            let moved_current = self.lp_arena.add(IR::Invalid);
+            let projection = IR::SimpleProjection {
+                input: moved_current,
+                columns: proj_schema,
+            };
+            let current = self.lp_arena.replace(current, projection);
+            self.lp_arena.replace(moved_current, current);
+
+            // We know that this node is done optimizing
+            break;
+        }
+
+        let ir = self.lp_arena.get(current);
+        for child in ir.get_inputs().iter() {
```

Comment:
This pushes the nodes to a `Vec` stack every call. Maybe we can amortize the `Vec` above and call `copy_inputs`

We can then add an outer loop so that we aren't recursive anymore.